File: android/src/main/java/com/eskender/sae/sdk/models/Requests.kt
package com.eskender.sae.sdk.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.math.BigDecimal
import java.time.OffsetDateTime
import java.util.Date

@JsonClass(generateAdapter = true)
data class GenerarDocumentoRequest(
    @Json(name = "TipoDocumento") val tipoDocumento: String,
    @Json(name = "Clave") val clave: ClaveRequest,
    @Json(name = "Consecutivo") val consecutivo: ConsecutivoRequest,
    @Json(name = "CodigoActividadEmisor") val codigoActividadEmisor: String,
    @Json(name = "CodigoActividadReceptor") val codigoActividadReceptor: String,
    @Json(name = "FechaEmision") val fechaEmision: String, // String for simplicity, or use Adapter
    @Json(name = "ProveedorSistemas") val proveedorSistemas: String,
    @Json(name = "Emisor") val emisor: EmisorRequest,
    @Json(name = "Receptor") val receptor: ReceptorRequest,
    @Json(name = "CondicionVenta") val condicionVenta: String,
    @Json(name = "CondicionVentaOtros") val condicionVentaOtros: CondicionVentaOtro? = null,
    @Json(name = "PlazoCredito") val plazoCredito: Int? = null,
    @Json(name = "DetalleServicio") val detalleServicio: DetalleServicioRequest,
    @Json(name = "ResumenFactura") val resumenFactura: ResumenFacturaRequest?,
    @Json(name = "InformacionReferencia") val informacionReferencia: List<InformacionReferenciaRequest>? = null,
    @Json(name = "OtrosCargos") val otrosCargos: List<OtrosCargosRequest>? = null,
    @Json(name = "OtrosDatos") val otrosDatos: OtrosRequest? = null
)

@JsonClass(generateAdapter = true)
data class ClaveRequest(
    @Json(name = "CodigoPais") val codigoPais: Int,
    @Json(name = "Dia") val dia: Int,
    @Json(name = "Mes") val mes: Int,
    @Json(name = "Ano") val ano: Int,
    @Json(name = "Consecutivo") val consecutivo: ConsecutivoRequest,
    @Json(name = "CedulaEmisor") val cedulaEmisor: String,
    @Json(name = "Situacion") val situacion: String
)

@JsonClass(generateAdapter = true)
data class ConsecutivoRequest(
    @Json(name = "Establecimiento") val establecimiento: Int,
    @Json(name = "Terminal") val terminal: Int,
    @Json(name = "NumeroConsecutivo") val numeroConsecutivo: Long,
    @Json(name = "TipoComprobante") val tipoComprobante: String
)

@JsonClass(generateAdapter = true)
data class IdentificacionRequest(
    @Json(name = "Tipo") val tipo: String,
    @Json(name = "Numero") val numero: String
)

@JsonClass(generateAdapter = true)
data class EmisorRequest(
    @Json(name = "Identificacion") val identificacion: IdentificacionRequest,
    @Json(name = "Nombre") val nombre: String,
    @Json(name = "NombreComercial") val nombreComercial: String? = null,
    @Json(name = "Ubicacion") val ubicacion: UbicacionRequest,
    @Json(name = "Telefono") val telefono: TelefonoRequest,
    @Json(name = "CorreoElectronico") val correoElectronico: List<String>,
    @Json(name = "RegistroFiscal8707") val registroFiscal8707: String? = null,
    @Json(name = "SeFacturanBebidasAlc") val seFacturanBebidasAlc: Boolean = false
)

@JsonClass(generateAdapter = true)
data class ReceptorRequest(
    @Json(name = "Identificacion") val identificacion: IdentificacionRequest,
    @Json(name = "Nombre") val nombre: String,
    @Json(name = "IdentificacionExtranjero") val identificacionExtranjero: String? = null,
    @Json(name = "NombreComercial") val nombreComercial: String? = null,
    @Json(name = "Ubicacion") val ubicacion: UbicacionRequest? = null,
    @Json(name = "OtrasSenasExtranjero") val otrasSenasExtranjero: String? = null,
    @Json(name = "Telefono") val telefono: TelefonoRequest? = null,
    @Json(name = "CorreoElectronico") val correoElectronico: String // Single string in C# Request
)

@JsonClass(generateAdapter = true)
data class TelefonoRequest(
    @Json(name = "CodigoPais") val codigoPais: Int,
    @Json(name = "Numero") val numero: Long
)

@JsonClass(generateAdapter = true)
data class UbicacionRequest(
    @Json(name = "Provincia") val provincia: String,
    @Json(name = "Canton") val canton: String,
    @Json(name = "Distrito") val distrito: String,
    @Json(name = "Barrio") val barrio: String,
    @Json(name = "OtrasSenas") val otrasSenas: String
)

@JsonClass(generateAdapter = true)
data class DetalleServicioRequest(
    @Json(name = "LineasDetalle") val lineasDetalle: List<LineaDetalleRequest> = emptyList()
)

@JsonClass(generateAdapter = true)
data class LineaDetalleRequest(
    @Json(name = "NumeroLinea") var numeroLinea: Int? = null,
    @Json(name = "PartidaArancelaria") val partidaArancelaria: String? = null,
    @Json(name = "CodigoCABYS") val codigoCabys: String? = null,
    @Json(name = "CodigosComerciales") val codigosComerciales: List<CodigoComercialRequest> = emptyList(),
    @Json(name = "Cantidad") val cantidad: BigDecimal,
    @Json(name = "UnidadMedida") val unidadMedida: String,
    @Json(name = "TipoTransaccion") val tipoTransaccion: String,
    @Json(name = "UnidadMedidaComercial") val unidadMedidaComercial: String? = null,
    @Json(name = "Detalle") val detalle: String? = null,
    @Json(name = "NumeroVINoSerie") val numeroVinOSerie: List<String>? = null,
    @Json(name = "RegistroMedicamento") val registroMedicamento: String? = null,
    @Json(name = "FormaFarmaceutica") val formaFarmaceutica: String? = null,
    // @Json(name = "DetalleSurtido") val detalleSurtido: DetalleSurtidoRequest? = null, // Skipping specific complex types for now if not used
    @Json(name = "PrecioUnitario") val precioUnitario: BigDecimal,
    @Json(name = "MontoTotal") val montoTotal: BigDecimal? = null,
    @Json(name = "Descuentos") val descuentos: List<DescuentoRequest> = emptyList(),
    @Json(name = "SubTotal") val subTotal: BigDecimal? = null,
    @Json(name = "IVACobradoFabrica") val ivaCobradoFabrica: String? = null,
    @Json(name = "BaseImponible") val baseImponible: BigDecimal? = null,
    @Json(name = "Impuestos") val impuestos: List<ImpuestoRequest> = emptyList(),
    @Json(name = "ImpuestoAsumidoEmisorFabrica") val impuestoAsumidoEmisorFabrica: BigDecimal? = null,
    @Json(name = "ImpuestoNeto") val impuestoNeto: BigDecimal? = null,
    @Json(name = "MontoTotalLinea") val montoTotalLinea: BigDecimal? = null
)

@JsonClass(generateAdapter = true)
data class CodigoComercialRequest(
    @Json(name = "Tipo") val tipo: String,
    @Json(name = "Codigo") val codigo: String
)

@JsonClass(generateAdapter = true)
data class DescuentoRequest(
    @Json(name = "MontoDescuento") val montoDescuento: BigDecimal?,
    @Json(name = "CodigoDescuento") val codigoDescuento: String?,
    @Json(name = "CodigoDescuentoOtro") val codigoDescuentoOtro: String?,
    @Json(name = "NaturalezaDescuento") val naturalezaDescuento: String?
)

@JsonClass(generateAdapter = true)
data class ImpuestoRequest(
    @Json(name = "Codigo") val codigo: String,
    @Json(name = "CodigoImpuestoOTRO") val codigoImpuestoOtro: String?,
    @Json(name = "CodigoTarifaIVA") val codigoTarifaIva: String,
    @Json(name = "Tarifa") val tarifa: BigDecimal?,
    @Json(name = "FactorCalculoIVA") val factorCalculoIva: BigDecimal?,
    @Json(name = "Monto") val monto: BigDecimal?,
    @Json(name = "Exoneracion") val exoneracion: ExoneracionRequest?
)

@JsonClass(generateAdapter = true)
data class ExoneracionRequest(
    @Json(name = "TipoDocumentoEX1") val tipoDocumentoEx1: String,
    @Json(name = "TipoDocumentoOTRO") val tipoDocumentoOtro: String?,
    @Json(name = "NumeroDocumento") val numeroDocumento: String,
    @Json(name = "Articulo") val articulo: Int,
    @Json(name = "Inciso") val inciso: Int,
    @Json(name = "NombreInstitucion") val nombreInstitucion: String,
    @Json(name = "NombreInstitucionOtros") val nombreInstitucionOtros: String?,
    @Json(name = "FechaEmisionEX") val fechaEmisionEx: String,
    @Json(name = "TarifaExonerada") val tarifaExonerada: BigDecimal,
    @Json(name = "MontoExoneracion") val montoExoneracion: BigDecimal
)

@JsonClass(generateAdapter = true)
data class ResumenFacturaRequest(
    @Json(name = "CodigoTipoMoneda") val codigoTipoMoneda: CodigoTipoMonedaRequest?,
    @Json(name = "TotalServGravados") val totalServGravados: BigDecimal?,
    @Json(name = "TotalServExentos") val totalServExentos: BigDecimal?,
    @Json(name = "TotalServExonerado") val totalServExonerado: BigDecimal?,
    @Json(name = "TotalServNoSujeto") val totalServNoSujeto: BigDecimal?,
    @Json(name = "TotalMercanciasGravadas") val totalMercanciasGravadas: BigDecimal?,
    @Json(name = "TotalMercanciasExentas") val totalMercanciasExentas: BigDecimal?,
    @Json(name = "TotalMercExonerada") val totalMercExonerada: BigDecimal?,
    @Json(name = "TotalMercNoSujeta") val totalMercNoSujeta: BigDecimal?,
    @Json(name = "TotalGravado") val totalGravado: BigDecimal?,
    @Json(name = "TotalExento") val totalExento: BigDecimal?,
    @Json(name = "TotalExonerado") val totalExonerado: BigDecimal?,
    @Json(name = "TotalNoSujeto") val totalNoSujeto: BigDecimal?,
    @Json(name = "TotalVenta") val totalVenta: BigDecimal?,
    @Json(name = "TotalDescuentos") val totalDescuentos: BigDecimal?,
    @Json(name = "TotalVentaNeta") val totalVentaNeta: BigDecimal?,
    @Json(name = "TotalDesgloseImpuesto") val totalDesgloseImpuesto: TotalDesgloseImpuestoRequest?,
    @Json(name = "TotalImpuesto") val totalImpuesto: BigDecimal?,
    @Json(name = "TotalImpAsumEmisorFabrica") val totalImpAsumEmisorFabrica: BigDecimal?,
    @Json(name = "TotalIvaDevuelto") val totalIvaDevuelto: BigDecimal?,
    @Json(name = "TotalOtrosCargos") val totalOtrosCargos: BigDecimal?,
    @Json(name = "TipoMedioPago") val tipoMedioPago: MedioPagoRequest?,
    @Json(name = "TotalComprobante") val totalComprobante: BigDecimal?
)

@JsonClass(generateAdapter = true)
data class CodigoTipoMonedaRequest(
    @Json(name = "CodigoMoneda") val codigoMoneda: String,
    @Json(name = "TipoCambio") val tipoCambio: BigDecimal
)

@JsonClass(generateAdapter = true)
data class TotalDesgloseImpuestoRequest(
    @Json(name = "Codigo") val codigo: String,
    @Json(name = "CodigoTarifaIVA") val codigoTarifaIva: String,
    @Json(name = "TotalMontoImpuesto") val totalMontoImpuesto: BigDecimal
)

@JsonClass(generateAdapter = true)
data class MedioPagoRequest(
    @Json(name = "TipoMedioPago") val tipoMedioPago: String?,
    @Json(name = "MedioPagoOtros") val medioPagoOtros: String?,
    @Json(name = "TotalMedioPago") val totalMedioPago: BigDecimal?
)

@JsonClass(generateAdapter = true)
data class InformacionReferenciaRequest(
    @Json(name = "TipoDocIR") val tipoDocIr: String,
    @Json(name = "TipoDocRefOTRO") val tipoDocRefOtro: String?,
    @Json(name = "Numero") val numero: String,
    @Json(name = "FechaEmisionIR") val fechaEmisionIr: String,
    @Json(name = "Codigo") val codigo: String,
    @Json(name = "CodigoReferenciaOTRO") val codigoReferenciaOtro: String?,
    @Json(name = "Razon") val razon: String
)

@JsonClass(generateAdapter = true)
data class OtrosCargosRequest(
    @Json(name = "TipoDocumentoOC") val tipoDocumentoOc: String,
    @Json(name = "TipoDocumentoOTROS") val tipoDocumentoOtros: String?,
    @Json(name = "IdentificacionTercero") val identificacionTercero: IdentificacionRequest,
    @Json(name = "NombreTercero") val nombreTercero: String,
    @Json(name = "Detalle") val detalle: String,
    @Json(name = "PorcentajeOC") val porcentajeOc: BigDecimal,
    @Json(name = "MontoCargo") val montoCargo: BigDecimal
)

@JsonClass(generateAdapter = true)
data class OtrosRequest(
    @Json(name = "OtroTexto") val otroTexto: List<String>?,
    @Json(name = "OtroContenido") val otroContenido: String?
)

@JsonClass(generateAdapter = true)
data class CondicionVentaOtro(
    @Json(name = "Descripcion") val descripcion: String
)

@JsonClass(generateAdapter = true)
data class EnviarDocumentoSimplificadoRequest(
    @Json(name = "TipoDocumento") val tipoDocumento: String = "01",
    @Json(name = "CondicionVenta") val condicionVenta: String = "01",
    @Json(name = "MedioPago") val medioPago: List<String> = listOf("01"),
    @Json(name = "MediosPago") val mediosPago: List<MedioPagoRequest>? = null,
    @Json(name = "CodigoMoneda") val codigoMoneda: String = "CRC",
    @Json(name = "TipoCambio") val tipoCambio: BigDecimal = BigDecimal.ONE,
    @Json(name = "PlazoCredito") val plazoCredito: Int? = null,
    @Json(name = "Receptor") val receptor: ReceptorRequest? = null,
    @Json(name = "Lineas") val lineas: List<LineaDetalleRequest> = emptyList(),
    @Json(name = "Referencia") val referencia: ReferenciaSimplificada? = null,
    @Json(name = "OtrosCargos") val otrosCargos: List<OtrosCargosRequest>? = null
)

@JsonClass(generateAdapter = true)
data class ReferenciaSimplificada(
    @Json(name = "ClaveDocumentoReferencia") val claveDocumentoReferencia: String,
    @Json(name = "Codigo") val codigo: String = "01",
    @Json(name = "Razon") val razon: String
)


@JsonClass(generateAdapter = true)
data class InviteUserRequest(@Json(name = "Email") val email: String, @Json(name = "Role") val role: String)

@JsonClass(generateAdapter = true)
data class HaciendaConfigRequest(
    @Json(name = "IdCia") val idCia: String? = null,
    @Json(name = "IdSucursal") val idSucursal: String? = null,
    @Json(name = "PinP12") val pinP12: String? = null,
    @Json(name = "CertP12Base64") val certP12Base64: String? = null,
    @Json(name = "UsuarioOauth") val usuarioOauth: String? = null,
    @Json(name = "PasswordOauth") val passwordOauth: String? = null,
    @Json(name = "Environment") val environment: String = "Sandbox"
)

@JsonClass(generateAdapter = true)
data class MensajeEndosoRequest(
    @Json(name = "Clave") val clave: String,
    @Json(name = "NumeroCedulaEmisor") val numeroCedulaEmisor: String,
    @Json(name = "FechaEmisionDoc") val fechaEmisionDoc: String, // ISO String
    @Json(name = "DetalleMensaje") val detalleMensaje: String,
    @Json(name = "MontoTotalImpuesto") val montoTotalImpuesto: BigDecimal,
    @Json(name = "TotalFactura") val totalFactura: BigDecimal,
    @Json(name = "NumeroCedulaReceptor") val numeroCedulaReceptor: String,
    @Json(name = "NumeroConsecutivoReceptor") val numeroConsecutivoReceptor: String,
    @Json(name = "Mensaje") val mensaje: Int = 4,
    @Json(name = "NombreReceptor") val nombreReceptor: String? = null,
    @Json(name = "TipoIdentificacionReceptor") val tipoIdentificacionReceptor: String? = null
)

File: typescript/src/models/requests.ts
import {
    TipoComprobante,
    TipoIdentificacion,
    CondicionVenta,
    TipoMedioPago,
    CodigoMoneda,
    CodigoImpuesto,
    HaciendaEstado,
    LicenseType,
    LicensePlatform
} from './enums';

export interface CodigoComercialRequest {
    tipo: string;
    codigo: string;
}

export interface GenerarDocumentoRequest {
    tipoDocumento: string;
    clave: ClaveRequest;
    consecutivo: ConsecutivoRequest;
    codigoActividadEmisor: string;
    codigoActividadReceptor: string;
    fechaEmision: string; // ISO Date
    proveedorSistemas: string;
    emisor: EmisorRequest;
    receptor: ReceptorRequest;
    condicionVenta: string;
    condicionVentaOtros?: CondicionVentaOtro;
    plazoCredito?: number;
    detalleServicio: DetalleServicioRequest;
    resumenFactura?: ResumenFacturaRequest;
    informacionReferencia?: InformacionReferenciaRequest[];
    otrosCargos?: OtrosCargosRequest[];
    otrosDatos?: OtrosRequest;
}

export interface ClaveRequest {
    codigoPais: number;
    dia: number;
    mes: number;
    ano: number;
    consecutivo: ConsecutivoRequest;
    cedulaEmisor: string;
    situacion: string;
}

export interface ConsecutivoRequest {
    establecimiento: number;
    terminal: number;
    numeroConsecutivo: number;
    tipoComprobante: string;
}

export interface IdentificacionRequest {
    tipo: string;
    numero: string;
}

export interface EmisorRequest {
    identificacion: IdentificacionRequest;
    nombre: string;
    nombreComercial?: string;
    ubicacion: UbicacionRequest;
    telefono: TelefonoRequest;
    correoElectronico: string[];
    registroFiscal8707?: string;
    seFacturanBebidasAlc: boolean;
}

export interface TelefonoRequest {
    codigoPais: number;
    numero: number;
}

export interface ReceptorRequest {
    identificacion: IdentificacionRequest;
    nombre: string;
    identificacionExtranjero?: string;
    nombreComercial?: string;
    ubicacion?: UbicacionRequest;
    otrasSenasExtranjero?: string;
    telefono?: TelefonoRequest;
    correoElectronico: string;
}

export interface UbicacionRequest {
    provincia: string;
    canton: string;
    distrito: string;
    barrio: string;
    otrasSenas: string;
}

export interface DetalleServicioRequest {
    lineasDetalle: LineaDetalleRequest[];
}

export interface LineaDetalleRequest {
    numeroLinea?: number;
    partidaArancelaria?: string;
    codigoCABYS?: string;
    codigosComerciales: CodigoComercialRequest[];
    cantidad: number;
    unidadMedida: string;
    tipoTransaccion: string;
    unidadMedidaComercial?: string;
    detalle?: string;
    numeroVINoSerie?: string[];
    registroMedicamento?: string;
    formaFarmaceutica?: string;
    precioUnitario: number;
    montoTotal?: number;
    descuentos: DescuentoRequest[];
    subTotal?: number;
    ivaCobradoFabrica?: string;
    baseImponible?: number;
    impuestos: ImpuestoRequest[];
    impuestoAsumidoEmisorFabrica?: number;
    impuestoNeto?: number;
    montoTotalLinea?: number;
}

export interface DescuentoRequest {
    montoDescuento?: number;
    codigoDescuento?: string;
    codigoDescuentoOtro?: string;
    naturalezaDescuento?: string;
}

export interface ImpuestoRequest {
    codigo: string;
    codigoImpuestoOTRO?: string;
    codigoTarifaIVA: string;
    tarifa?: number;
    factorCalculoIVA?: number;
    monto?: number;
    exoneracion?: ExoneracionRequest;
}

export interface ExoneracionRequest {
    tipoDocumentoEX1: string;
    tipoDocumentoOTRO?: string;
    numeroDocumento: string;
    articulo: number;
    inciso: number;
    nombreInstitucion: string;
    nombreInstitucionOtros?: string;
    fechaEmisionEX: string; // ISO Date or DateOffset
    tarifaExonerada: number;
    montoExoneracion: number;
}

export interface ResumenFacturaRequest {
    codigoTipoMoneda?: CodigoTipoMonedaRequest;
    totalServGravados?: number;
    totalServExentos?: number;
    totalServExonerado?: number;
    totalServNoSujeto?: number;
    totalMercanciasGravadas?: number;
    totalMercanciasExentas?: number;
    totalMercExonerada?: number;
    totalMercNoSujeta?: number;
    totalGravado?: number;
    totalExento?: number;
    totalExonerado?: number;
    totalNoSujeto?: number;
    totalVenta?: number;
    totalDescuentos?: number;
    totalVentaNeta?: number;
    totalDesgloseImpuesto?: TotalDesgloseImpuestoRequest;
    totalImpuesto?: number;
    totalImpAsumEmisorFabrica?: number;
    totalIvaDevuelto?: number;
    totalOtrosCargos?: number;
    tipoMedioPago?: MedioPagoRequest;
    totalComprobante?: number;
}

export interface CodigoTipoMonedaRequest {
    codigoMoneda: string;
    tipoCambio: number;
}

export interface TotalDesgloseImpuestoRequest {
    codigo: string;
    codigoTarifaIVA: string;
    totalMontoImpuesto: number;
}

export interface MedioPagoRequest {
    tipoMedioPago?: string;
    medioPagoOtros?: string;
    totalMedioPago?: number;
}

export interface InformacionReferenciaRequest {
    tipoDocIR: string;
    tipoDocRefOTRO?: string;
    numero: string;
    fechaEmisionIR: string;
    codigo: string;
    codigoReferenciaOTRO?: string;
    razon: string;
}

export interface OtrosCargosRequest {
    tipoDocumentoOC: string;
    tipoDocumentoOTROS?: string;
    identificacionTercero: IdentificacionRequest;
    nombreTercero: string;
    detalle: string;
    porcentajeOC: number;
    montoCargo: number;
}

export interface OtrosRequest {
    otroTexto?: string[];
    otroContenido?: string;
}

export interface CondicionVentaOtro {
    descripcion: string;
}

/**
 * DTO simplificado para enviar documentos electrónicos.
 */
export interface EnviarDocumentoSimplificadoRequest {
    /** 01=FE, 02=ND, 03=NC, 04=Tiquete, 08=FEC, 09=FECE */
    tipoDocumento?: string;
    /** 01=Contado, 02=Crédito, etc. */
    condicionVenta?: string;
    /** Medios de pago: 01=Efectivo, 02=Tarjeta, etc. */
    medioPago?: string[];
    /** Lista detallada de medios de pago para facturas multipartes. */
    mediosPago?: MedioPagoRequest[];
    /** Código de moneda ISO (CRC, USD, EUR) */
    codigoMoneda?: string;
    /** Tipo de cambio respecto a CRC (1 si es CRC) */
    tipoCambio?: number;
    /** Plazo de crédito en días (solo si CondicionVenta=02) */
    plazoCredito?: number;
    /** Receptor/Cliente. Null = Consumidor Final / Sin Receptor. */
    receptor?: ReceptorRequest;
    /** Líneas de detalle del documento (requerido, mínimo 1) */
    lineas: LineaDetalleRequest[];
    /** Referencia para Notas de Crédito/Débito. */
    referencia?: ReferenciaSimplificada;
    /** Otros cargos (opcional) */
    otrosCargos?: OtrosCargosRequest[];
    /** Datos adicionales (opcional) */
    otrosDatos?: OtrosRequest;
    /** Código de actividad económica del emisor (opcional) */
    codigoActividadEmisor?: string;
    /** Código de actividad económica del receptor (opcional) */
    codigoActividadReceptor?: string;
}

/**
 * Referencia simplificada para NC/ND.
 */
export interface ReferenciaSimplificada {
    /** Clave numérica (50 dígitos) del documento que se referencia */
    claveDocumentoReferencia: string;
    /** Código de referencia Hacienda: 01=Anula, 02=Corrige texto, etc. */
    codigo?: string;
    /** Razón de la nota (descripción libre) */
    razon: string;
}

// --- TENANT ---

export interface UpdateTenantRequest {
    name?: string;
    phone?: string;
    taxId?: string;
    ownerEmail?: string;
    identificationType?: number;
    commercialName?: string;
    provinceCode?: string;
    cantonCode?: string;
    districtCode?: string;
    addressDetails?: string;
    economicActivityCode?: string;
    isActive?: boolean;
}

export interface RegisterTenantRequest {
    name: string;
    ownerEmail: string;
    taxId: string;
}

export interface HaciendaConfigRequest {
    idCia?: string;
    idSucursal?: string;
    pinP12?: string;
    certP12Base64?: string;
    usuarioOauth?: string;
    passwordOauth?: string;
    environment?: string; // "Sandbox" | "Production"
}

export interface SwitchEnvironmentRequest {
    environment: number | string;
}

export interface CreateBranchRequest {
    name: string;
    code: string;
    address?: string | null;
    phone?: string | null;
    managerName?: string | null;
}

export interface UpdateBranchRequest {
    name?: string | null;
    address?: string | null;
    phone?: string | null;
    managerName?: string | null;
    isActive?: boolean | null;
}

export interface CreateTerminalRequest {
    name: string;
    code: string;
}

export interface UpdateTerminalRequest {
    name?: string;
    isActive?: boolean;
}

export interface UpdateSequenceRequest {
    newCurrentValue: number;
}

// --- AUTH ---

export interface LoginRequest {
    username: string;
    password?: string;
    rememberMe?: boolean;
}

export interface ForgotPasswordRequest {
    email: string;
}

export interface ConfirmEmailRequest {
    email: string;
    token: string;
}

export interface ResetPasswordRequest {
    token: string;
    newPassword: string;
}

export interface RegisterRequest {
    companyName: string;
    taxId: string;
    email: string;
    password: string;
    tier?: number;
    sellerCode?: string;
}


export interface PartnerRegisterRequest {
    code: string;
    phoneNumber?: string;
    website?: string;
    bankName?: string;
    bankAccount?: string;
    taxId?: string;
}

export interface IssueLicenseRequest {
    type: LicenseType;
    platform: LicensePlatform | string | number;
    deviceName?: string;
    sourceAddonId?: string;
}

export interface LeadRegisterRequest {
    name: string;
    email: string;
    phone?: string;
    message?: string;
    source?: string;
}

export interface TenantRoleRequest {
    name: string;
    description: string;
    permissions: string[];
}

export interface MensajeEndosoRequest {
    clave: string;
    numeroCedulaEmisor: string;
    fechaEmisionDoc: string; // ISO Date
    detalleMensaje: string;
    montoTotalImpuesto: number;
    totalFactura: number;
    numeroCedulaReceptor: string;
    numeroConsecutivoReceptor: string;
    mensaje: number; // 1: Aceptado, 2: Parcial, 3: Rechazo, 4: Cesión
    nombreReceptor?: string;
    tipoIdentificacionReceptor?: string;
}

export interface CreateSettlementRequest {
    commissionIds: string[];
}

export interface RegisterSettlementPaymentRequest {
    paymentReference: string;
    paymentDate: string; // ISO
}

// --- LICENSES ---

export interface ValidateLicenseRequest {
    licenseKey: string;
    machineId: string;
    appVersion?: string;
    type?: number | string; // LicenseType enum
    platform?: number | string; // LicensePlatform enum
    appIdentifier?: string;
}

export interface ActivateLicenseRequest {
    licenseKey: string;
    machineId: string;
    deviceName?: string;
    type?: number | string;
    platform?: number | string;
    appIdentifier?: string;
}

export interface PurchaseAddonRequest {
    packageId: string;
}

export interface ReportAddonPurchaseRequest {
    packageId: string;
    reference: string;
    paymentMethod: string; // Use enum if available, or string
    amount: number;
}

export interface ApproveAddonRequest {
    adminNote?: string;
}

export interface RejectAddonRequest {
    reason: string;
}

export interface CreateLicensePackageRequest {
    name: string;
    description: string;
    type: string | number; // LicenseType enum value
    allowedPlatforms: (string | number)[]; // LicensePlatform enum value array
    quantity: number;
    monthlyPrice: number;
    currency?: string;
    isActive?: boolean;
    displayOrder?: number;
    defaultModules?: string | null;
    iconUrl?: string | null;
    appIdentifier?: string | null;
}

export interface UpdateLicensePackageRequest {
    name?: string;
    description?: string;
    type?: string | number;
    allowedPlatforms?: (string | number)[];
    quantity?: number;
    monthlyPrice?: number;
    currency?: string;
    isActive?: boolean;
    displayOrder?: number;
    defaultModules?: string | null;
    iconUrl?: string | null;
    appIdentifier?: string | null;
}

export interface RegisterAppRequest {
    appName: string;
    appSlug: string;
    description?: string;
    websiteUrl?: string | null;
    requiredType: string | number;
    requiredPlatform: string | number;
}

export interface CreateOrderRequest {
    items: CartItemDto[];
    sellerCode?: string;
}

export interface CartItemDto {
    type: number;
    id?: string;
    quantity: number;
}

export interface ReportOrderPaymentRequest {
    orderId: string;
    referenceNumber: string;
    paymentMethod: number;
}

export interface InviteUserRequest {
    email: string;
    role: number;
    customRoleId?: string | null;
    branchId?: string | null;
}

export interface UpdateUserRoleRequest {
    role: number;
    customRoleId?: string | null;
}

export interface AcceptInviteRequest {
    token: string;
    password?: string;
    fullName?: string;
}

export interface AdminRegisterRequest {
    email: string;
    password?: string;
    fullName: string;
    role: string;
}

export interface AdminUpdateRequest {
    email?: string;
    fullName?: string;
    role?: string;
    isActive?: boolean;
}

export interface ApprovalRejectRequest {
    reason: string;
}

export interface SettlementPayRequest {
    paymentReference: string;
    paymentDate: string;
}

// --- MODULES ---

export interface CreatePlatformModuleRequest {
    code: string;
    name: string;
    description: string;
    type: number; // ModuleType
    icon?: string;
    isActive: boolean;
    appIdentifier?: string;
}

export interface UpdatePlatformModuleRequest {
    name?: string;
    description?: string;
    type?: number;
    icon?: string;
    isActive?: boolean;
    appIdentifier?: string;
}
// --- DEVELOPER MODULES ---

export interface CreateDeveloperModuleRequest {
    code: string;
    name: string;
    description: string;
    type: number;
    isActive: boolean;
    appIdentifier?: string | null;
}

export interface UpdateDeveloperModuleRequest {
    name?: string;
    description?: string;
    type?: number;
    isActive?: boolean;
    appIdentifier?: string | null;
}
